src/Shared/ServerInfrastructure/BufferExtensions.cs

```diff
@@ -129,47 +129,43 @@ internal static void WriteAscii(ref this BufferWriter<PipeWriter> buffer, string
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    internal static unsafe void WriteNumeric(ref this BufferWriter<PipeWriter> buffer, ulong number)
+    internal static void WriteNumeric(ref this BufferWriter<PipeWriter> buffer, ulong number)
     {
         const byte AsciiDigitStart = (byte)'0';
 
-        var span = buffer.Span;
-        var bytesLeftInBlock = span.Length;
+        var start = buffer.Span;
+        var bytesLeftInBlock = start.Length;
 
         // Fast path, try copying to the available memory directly
         var simpleWrite = true;
-        fixed (byte* output = span)
+        if (number < 10 && bytesLeftInBlock >= 1)
         {
-            var start = output;
-            if (number < 10 && bytesLeftInBlock >= 1)
-            {
-                *(start) = (byte)(((uint)number) + AsciiDigitStart);
-                buffer.Advance(1);
-            }
-            else if (number < 100 && bytesLeftInBlock >= 2)
-            {
-                var val = (uint)number;
-                var tens = (byte)((val * 205u) >> 11); // div10, valid to 1028
+            start[0] = (byte)(((uint)number) + AsciiDigitStart);
+            buffer.Advance(1);
+        }
+        else if (number < 100 && bytesLeftInBlock >= 2)
+        {
+            var val = (uint)number;
+            var tens = (byte)((val * 205u) >> 11); // div10, valid to 1028
```

Comment:
```suggestion
            var tens = (uint)(byte)((val * 205u) >> 11); // div10, valid to 1028
```
Avoids a sign extending move.
But is the trick for `div10, valid to 1028` still valid then? I think so, as the `byte` cast isn't removed.

This could be done at the other places too.

Reply:
Is the suggested code what you had in mind? If we would want to avoid the signed shift, should it be, no? 

```csharp
var tens = (byte)(((uint)(val * 205u)) >> 11); // div10, valid to 1028
```

But `val` is already `uint`. I see it is already using `SHR` (instead of `SAR`) on x64:

```
imul      eax,r8d,0CD
shr       eax,0B
```

Or did you mean something else?

Reply:
Not about the shift, later on there should be a `movsxd` that can be avoided.

ATM and w/ the code from your comment the type of `tens` is `byte`.
So there's a sign extending move later on.

W/ the explicit cast to `uint` (or when not `var` is used, but `uint` as type 😉) instead of the sign extension the faster zero extending move can be used.

But all of this only happens when my brain compiles the code correctly (and unfortunately recent JIT improvements sometimes out-perform my brain compiler...).

Reply:
I can see the `movsxd` in the old 'unsafe' code, but only `movzx` for the jitted Span based one. Here is the block for the for 3 digits branch:


```
      cmp       r8,3E8
       jae       short M00_L03
       cmp       eax,3
       jl        short M00_L03
       imul      eax,r8d,29
       shr       eax,0C
       movzx     ecx,al
       imul      eax,r8d,0CD
       shr       eax,0B
       movzx     eax,al
       lea       r10d,[rcx+30]
       mov       [rdx],r10b
       lea       ecx,[rcx+rcx*4]
       add       ecx,ecx
       mov       r10d,eax
       sub       r10d,ecx
       add       r10d,30
       mov       [rdx+1],r10b
       mov       ecx,r8d
       lea       eax,[rax+rax*4]
       add       eax,eax
       cdqe
       sub       rcx,rax
       add       rcx,30
       mov       [rdx+2],cl
M00_L03:
       ret
```

movzx is zero extending, which is fine, but I thi `cdqe` is the sign extending. With `uint` though the very end differs, there is no `cdge`

```
       mov       [rdx+1],r10b
       lea       eax,[rax+rax*4]
       add       eax,eax
       sub       r8d,eax
       add       r8d,30
       mov       [rdx+2],r8b
M00_L03:
       ret
```

I update the PR as per the suggestion.

Reply:
Updated the PR as suggested. 

@gfoidl which source (document, webpage, etc.) do you use for latency and throughput for these instructions?

Reply:
https://uops.info/

But for the sign extending moves I didn't look, because they're slower (than zero extending ones) since these moves exists 😉.

Reply:
thank you

Reply:
I will mark this as resolved